server/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java

```diff
@@ -461,18 +462,39 @@ private void restore(final IndexShard indexShard, final Repository repository, f
-            assert indexShard.shardRouting.primary() : "only primary shards can recover from store";
-            indexShard.openEngineAndRecoverFromTranslog();
-            indexShard.getEngine().fillSeqNoGaps(indexShard.getPendingPrimaryTerm());
+            try (Translog.Snapshot historySnapshot = repository.restoreShard(indexShard, restoreSource.snapshot().getSnapshotId(),
+                restoreSource.version(), indexId, snapshotShardId, indexShard.recoveryState())) {
+                final Store store = indexShard.store();
+                store.bootstrapNewHistory();
+                final SegmentInfos segmentInfos = store.readLastCommittedSegmentsInfo();
+                SequenceNumbers.CommitInfo commitInfo = SequenceNumbers.loadSeqNoInfoFromLuceneCommit(segmentInfos.userData.entrySet());
+                final String translogUUID = Translog.createEmptyTranslog(
+                    indexShard.shardPath().resolveTranslog(), commitInfo.maxSeqNo, shardId, indexShard.getPendingPrimaryTerm());
+                store.associateIndexWithNewTranslog(translogUUID);
+                assert indexShard.shardRouting.primary() : "only primary shards can recover from store";
+                indexShard.openEngineAndRecoverFromTranslog();
+                if (commitInfo.localCheckpoint < commitInfo.maxSeqNo) {
+                    Translog.Operation op;
+                    translogState.totalOperations(historySnapshot.totalOperations());
+                    while ((op = historySnapshot.next()) != null) {
+                        indexShard.applyTranslogOperation(op, Engine.Operation.Origin.SNAPSHOT_RECOVERY);
+                        translogState.incrementRecoveredOperations();
+                    }
+                    indexShard.getEngine().fillSeqNoGaps(indexShard.getPendingPrimaryTerm());
+                    /*
+                     * If there're gaps in the restoring commit, we won't have every translog entry from the local_checkpoint
+                     * to the max_seq_no. Then we won't have a safe commit for the restoring commit is not safe (missing translog).
+                     * To maintain the safe commit assumption, we have to forcefully flush a new commit here.
+                     */
+                    indexShard.flush(new FlushRequest().force(true).waitIfOngoing(true));
+                    final SequenceNumbers.CommitInfo newCommitInfo = SequenceNumbers.loadSeqNoInfoFromLuceneCommit(
+                        store.readLastCommittedSegmentsInfo().userData.entrySet());
+                    if (newCommitInfo.localCheckpoint < newCommitInfo.maxSeqNo) {
+                        throw new IndexShardRestoreFailedException(shardId, "history is not restored " +
+                            "checkpoint=" + newCommitInfo.localCheckpoint + " max_seq_no=" + newCommitInfo.maxSeqNo);
+                    }
+                }
+            }
             indexShard.finalizeRecovery();
             indexShard.postRecovery("restore done");
```

Comment:
should we assert in post_recovery that we have a safe commit?

Reply:
I added an assertion here, but this assertion may slow down our tests. I will post the difference.

Reply:
Sadly, we don't have this invariant with closed indices because the global checkpoint is not persisted to the translog checkpoint during recovery.